### PR TITLE
GH-35461: [C++] Add function Seek to BufferOutputStream

### LINKED ISSUE #35461: [C++] Add function Seek to BufferOutputStream
### Describe the enhancement requested

Add function Seek to BufferOutputStream

### Component(s)

C++

### PR DESCRIPTION
**Pull request by jinchengchenghh**
* Closes: #35461

### CODE REVIEW

File: cpp/src/arrow/io/memory.cc [hunk 1]
```diff
@@ -110,6 +110,21 @@ Status BufferOutputStream::Write(const void* data, int64_t nbytes) {
   return Status::OK();
 }
 
+Status BufferOutputStream::Seek(int64_t position) {
```
Comment (wgtmac) [inline cpp/src/arrow/io/memory.cc:113]: I'd prefer creating a separate class for the seekable output stream as its semantics is append-only and downstream applications are using it with that assumption.

Comment (jinchengchenghh) [reply cpp/src/arrow/io/memory.cc:113]: Do you have recommended class name for seekable output stream?

Comment (wgtmac) [reply cpp/src/arrow/io/memory.cc:113]: Before discussing the name, could you please justify the use case? 

Also cc @wjones127

Comment (westonpace) [reply cpp/src/arrow/io/memory.cc:113]: On the read side we have:

InputStream - non-seekable input
ReadableFile - seekable input

So I think we should use `WriteableFile` and not `OutputStream`.  Also note that some filesystems do not support seekable output (e.g. S3).

Comment (jinchengchenghh) [reply cpp/src/arrow/io/memory.cc:113]: I optimize my code to use `FixedSizeBufferWriter`, so I don't need this feature now, thanks for your kindly review!

**Pull request closed without merging**